homeassistant/components/roomba/vacuum.py

```diff
@@ -83,57 +54,52 @@
 SUPPORT_ROOMBA_CARPET_BOOST = SUPPORT_ROOMBA | SUPPORT_FAN_SPEED
 
 
-async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
-    """Set up the iRobot Roomba vacuum cleaner platform."""
-
-    if PLATFORM not in hass.data:
-        hass.data[PLATFORM] = {}
-
-    host = config.get(CONF_HOST)
-    name = config.get(CONF_NAME)
-    username = config.get(CONF_USERNAME)
-    password = config.get(CONF_PASSWORD)
-    certificate = config.get(CONF_CERT)
-    continuous = config.get(CONF_CONTINUOUS)
-    delay = config.get(CONF_DELAY)
-
-    roomba = Roomba(
-        address=host,
-        blid=username,
-        password=password,
-        cert_name=certificate,
-        continuous=continuous,
-        delay=delay,
-    )
-    _LOGGER.debug("Initializing communication with host %s", host)
-
-    try:
-        with async_timeout.timeout(9):
-            await hass.async_add_job(roomba.connect)
-    except asyncio.TimeoutError:
-        raise PlatformNotReady
-
-    roomba_vac = RoombaVacuum(name, roomba)
-    hass.data[PLATFORM][host] = roomba_vac
-
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up the iRobot Roomba vacuum cleaner."""
+    roomba = hass.data[DOMAIN][config_entry.entry_id]
+    roomba_vac = RoombaVacuum(roomba)
     async_add_entities([roomba_vac], True)
 
 
 class RoombaVacuum(VacuumDevice):
     """Representation of a Roomba Vacuum cleaner robot."""
 
-    def __init__(self, name, roomba):
+    def __init__(self, roomba):
         """Initialize the Roomba handler."""
         self._available = False
         self._battery_level = None
         self._capabilities = {}
         self._fan_speed = None
         self._is_on = False
-        self._name = name
         self._state_attrs = {}
         self._status = None
         self.vacuum = roomba
-        self.vacuum_state = None
+        self.vacuum_state = self.vacuum.master_state.get("state", {}).get(
+            "reported", {}
+        )
+        self._mac = self.vacuum_state.get("mac")
```

Comment:
> I did some testing with this and found it wasn't able to get the mac address
> 
> ```
> 020-04-04 16:08:13 ERROR (MainThread) [homeassistant.components.vacuum] Error while setting up roomba platform for vacuum
> Traceback (most recent call last):
>   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 186, in _async_setup_platform
>     await asyncio.gather(*pending)
>   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 295, in async_add_entities
>     await asyncio.gather(*tasks)
>   File "/usr/src/homeassistant/homeassistant/helpers/entity_platform.py", line 363, in _async_add_entity
>     device = device_registry.async_get_or_create(**processed_dev_info)
>   File "/usr/src/homeassistant/homeassistant/helpers/device_registry.py", line 122, in async_get_or_create
>     for key, value in connections
>   File "/usr/src/homeassistant/homeassistant/helpers/device_registry.py", line 122, in <setcomp>
>     for key, value in connections
>   File "/usr/src/homeassistant/homeassistant/helpers/device_registry.py", line 54, in format_mac
>     if len(to_test) == 17 and to_test.count(":") == 5:
> TypeError: object of type 'NoneType' has no len()
> ```
I tested the code and got the same error. It seems that the key `mac` does not exist in `vacuum_state`. However, I can get the correct mac address from `hwPartsRev/wlan0HwAddr` as shown below.
```suggestion
        self._mac = self.vacuum_state.get("hwPartsRev", {}).get("wlan0HwAddr")
```
I'm not sure if it works for all devices and software version. I only tested it on a Braava jet M6 with software version 3.6.6.

Reply:
mac exist , hwPartsRev not exist with Roomba 980
my master_state: 
> Got new state from the vacuum: OrderedDict([('netinfo', OrderedDict([('dhcp', True), ('addr', 323223XXXX), ('mask', 429496XXXX), ('gw', 323223XXXX), ('dns1', 3232235778), ('dns2', 134744072), ('bssid', '0e:36:c9:XX:XX:XX'), ('sec', 4)])), ('wifistat', OrderedDict([('wifi', 1), ('uap', False), ('cloud', 1)])), ('wlcfg', OrderedDict([('sec', 7), ('ssid', '6D696C6B79XXXXX')])), ('mac', 'd0:c5:d3:XX:XX:XX'), ('country', 'XX'), ('cloudEnv', 'prod'), ('svcEndpoints', OrderedDict([('svcDeplId', 'v011')])), ('mapUploadAllowed', True), ('localtimeoffset', 120), ('utctime', 1586105119), ('pose', OrderedDict([('theta', 0), ('point', OrderedDict([('x', 0), ('y', 0)]))])), ('batPct', 100), ('dock', OrderedDict([('known', False)])), ('bin', OrderedDict([('present', True), ('full', False)])), ('audio', OrderedDict([('active', False)])), ('cleanMissionStatus',...........

Reply:
Likely need a `find_mac` function. Ideally this is provided by the pypi package

Reply:
```python
self._mac = self.vacuum_state.get("hwPartsRev", {}).get("wlan0HwAddr")
```
this at all not a good thing, all state in future can be reduced to only required parameters, so there will be no such things like this one

Reply:
We agreed to replace the mac address with the blid so that it can be compatible